#**Exercise 1**

In [1]:
!pip install -q pyomo
from pyomo.environ import *

In [2]:
model=ConcreteModel()

In [3]:
model.x1 = Var(domain=Reals)
model.x2 = Var(domain=Reals)
model.x3 = Var(domain=Reals)
model.x4 = Var(domain=Reals)
model.x5 = Var(domain=Reals)
model.x6 = Var(domain=Reals)
model.x7 = Var(domain=Reals)

In [4]:
model.z = Objective(expr = 2*model.x1 - 5*model.x2 + 14*model.x3 + 7*model.x4 - 11*model.x5 + 0.17*model.x6 - 10.2*model.x7, sense=minimize)

In [5]:
model.constr1 = Constraint(expr = model.x1+model.x2+model.x3+model.x4+model.x5+model.x6+model.x7==280)
model.constr2 = Constraint(expr = model.x1-model.x2 >= 10)
model.constr3 = Constraint(expr = model.x6-3*model.x3-3*model.x7 <= 205)
model.constr4 = Constraint(expr = 20*model.x1+42*model.x6+model.x5 >= 35)
model.constr5 = Constraint(expr = model.x2-model.x3+4*model.x7 >= 24)
model.constr6 = Constraint(expr = model.x1 >= 12)
model.constr7 = Constraint(expr = model.x2 >= 0)
model.constr8 = Constraint(expr = model.x3 >= -5)
model.constr9 = Constraint(expr = model.x4 >= 3)
model.constr10= Constraint(expr = model.x5 >= 10)
model.constr11= Constraint(expr = model.x6 >= -4)
model.constr12= Constraint(expr = model.x7 >= 0)

In [6]:
!apt-get install -y -qq glpk-utils

In [7]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -3035.88
  Upper bound: -3035.88
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 26
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010447263717651367
# ----------------------------------------------------------
#   Solution Information
# ----------------------------

In [8]:
print('\nz = ', model.z())

print('\nDecision Variables')
print('x1 = ', model.x1())
print('x2 = ', model.x2())
print('x3 = ', model.x3())
print('x4 = ', model.x4())
print('x5 = ', model.x5())
print('x6 = ', model.x6())
print('x7 = ', model.x7())

print('\nConstraints')
print('Constraint 1  = ', model.constr1())
print('Constraint 2  = ', model.constr2())
print('Constraint 3  = ', model.constr3())
print('Constraint 4  = ', model.constr4())


z =  -3035.8799999999997

Decision Variables
x1 =  12.0
x2 =  0.0
x3 =  -5.0
x4 =  3.0
x5 =  269.25
x6 =  -4.0
x7 =  4.75

Constraints
Constraint 1  =  280.0
Constraint 2  =  12.0
Constraint 3  =  -3.25
Constraint 4  =  341.25


**Exercise 1 part 3** \\
The optimal value is ** -3035.88**
and the value of decision variables are as follows \\
$x_1$ = 12, $x_2$ = 0, $x_3$ = -5, $x_4$ = 3, $x_5$ = 269.25, $x_6$ = -4, $x_7$ = 4.75

**Exercise 1 part 4** \\
Activities of first four constraints respectively \\
Value of LHS followed by whether equality holds or not.

1.   280   , Equality holds
2.   12    , Equality does not holds
3.   -3.25 , Equality does not holds
4.   341.25, Equality does not holds





#**Exercise 2**

During the construction of following linear problem i have assumed that $x_1-x_2+x_3-x_4+x_5 \neq 0$ and also $x_1+x_2+x_3-x_5 \neq 0$ because as if then multiplication by $x_1-x_2+x_3-x_4+x_5$ and $x_1+x_2+x_3-x_5$ in inequality 3 and 4 respectively will not be possible.

So, the required linear problem is


$
\max \ 0.043x_1 + 0.027x_2 + 0.025x_3 + 0.022x_4 + 0.045x_5 - 0.0275y \\
\text{ s.t. } x_1+3x_2+x_3+7x_4+x_5-y \ \ \leq 100, \\
x_2+2x_3+4x_4 \ \ \geq 60, \\ 
0.8x_1 - 4.8x_2 + 1.8x_3 - 3.8x_4 - 2.2x_5 \geq 0, \\ 
x_1- 12x_2 - x_3 -3x_4 +5x_5 \geq 0 \\
y \geq 0 \\
y \leq 2.25 \\
x_1,x_2,x_3,x_4,x_5 \geq 0
$

In [9]:
model2= ConcreteModel()

In [10]:
model2.x1 = Var(domain=Reals)
model2.x2 = Var(domain=Reals)
model2.x3 = Var(domain=Reals)
model2.x4 = Var(domain=Reals)
model2.x5 = Var(domain=Reals)
model2.y  = Var(domain=Reals)

In [11]:
model2.z2 = Objective(expr = 0.043*model2.x1 + 0.027*model2.x2 + 0.025*model2.x3 + 0.022*model2.x4 + 0.045*model2.x5 - 0.0275*model2.y, sense = maximize)

In [12]:
model2.constr1 = Constraint(expr = model2.x1+3*model2.x2+model2.x3+7*model2.x4+model2.x5-model2.y <= 100)
model2.constr2 = Constraint(expr = model2.x2+2*model2.x3+4*model2.x4 >= 60)
model2.constr3 = Constraint(expr = 0.8*model2.x1-4.8*model2.x2+1.8*model2.x3-3.8*model2.x4-2.2*model2.x5 >= 0)
model2.constr4 = Constraint(expr = model2.x1-12*model2.x2-model2.x3-3*model2.x4+5*model2.x5 >= 0)
model2.constr5 = Constraint(expr = model2.y >= 0)
model2.constr6 = Constraint(expr = model2.y <= 2.25)
model2.constr7 = Constraint(expr = model2.x1 >= 0)
model2.constr8 = Constraint(expr = model2.x2 >= 0)
model2.constr9 = Constraint(expr = model2.x3 >= 0)
model2.constr10= Constraint(expr = model2.x4 >= 0)
model2.constr11= Constraint(expr = model2.x5 >= 0)


In [13]:
!apt-get install -y -qq glpk-utils

In [14]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 3.86940833333333
  Upper bound: 3.86940833333333
  Number of objectives: 1
  Number of constraints: 12
  Number of variables: 7
  Number of nonzeros: 27
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.014746427536010742
# ----------------------------------------------------------
#   Solution Information
# ------------

In [15]:
print('\nz2 = ', model2.z2())

print('\nDecision Variables')
print('x1 = ', model2.x1())
print('x2 = ', model2.x2())
print('x3 = ', model2.x3())
print('x4 = ', model2.x4())
print('x5 = ', model2.x5())
print('y = ', model2.y())

print('\nConstraints')
print('Constraint 1  = ', model2.constr1())
print('Constraint 2  = ', model2.constr2())
print('Constraint 3  = ', model2.constr3())
print('Constraint 4  = ', model2.constr4())
print('Constraint 5  = ', model2.constr5())
print('Constraint 6  = ', model2.constr6())
print('Constraint 7  = ', model2.constr7())
print('Constraint 8  = ', model2.constr8())
print('Constraint 9  = ', model2.constr9())
print('Constraint 10 = ', model2.constr10())
print('Constraint 11 = ', model2.constr11())


z2 =  3.869408333333333

Decision Variables
x1 =  34.9833333333333
x2 =  0.0
x3 =  30.0
x4 =  0.0
x5 =  37.2666666666667
y =  2.25

Constraints
Constraint 1  =  100.0
Constraint 2  =  60.0
Constraint 3  =  -1.1368683772161603e-13
Constraint 4  =  191.3166666666668
Constraint 5  =  2.25
Constraint 6  =  2.25
Constraint 7  =  34.9833333333333
Constraint 8  =  0.0
Constraint 9  =  30.0
Constraint 10 =  0.0
Constraint 11 =  37.2666666666667


**Exercise 2 part 3** \\
The optimal value is ** 3.87**
and the value of decision variables are as follows \\
$x_1$ = 34.983, $x_2$ = 0, $x_3$ = 30, $x_4$ = 0, $x_5$ = 37.267, $y$ = 2.25

Activities of all constraints respectively \\
Value of LHS followed by whether equality holds or not.

1.   100       , Equality holds
2.   60        , Equality holds
3.   0 (approx), Equality holds
4.   191.32    , Equality does not holds
5.   2.25      , Equality does not holds
6.   2.25      , Equality holds
7.   34.98     , Equality does not holds
8.   0         , Equality holds
9.   30        , Equality does not holds
10.  0         , Equality holds
11.  37.27     , Equality does not holds

**Exercise 2 part 4**



We can introduce following few more constraints then there will be no need of assumption. Which are \\
$x_1-x_2>=0$, $x_3-x_4>=0$, $x_3-x_5>=0$ \\
Then the modified linear program will become \\
$
\max \ 0.043x_1 + 0.027x_2 + 0.025x_3 + 0.022x_4 + 0.045x_5 - 0.0275y \\
\text{ s.t. } x_1+3x_2+x_3+7x_4+x_5-y \ \ \leq 100, \\
x_2+2x_3+4x_4 \ \ \geq 60, \\ 
0.8x_1 - 4.8x_2 + 1.8x_3 - 3.8x_4 - 2.2x_5 \geq 0, \\ 
x_1- 12x_2 - x_3 -3x_4 +5x_5 \geq 0 \\
y \geq 0 \\
y \leq 2.25 \\
x_1-x_2>=0 \\
x_3-x_4>=0 \\
x_3-x_5>=0 \\
x_1,x_2,x_3,x_4,x_5 \geq 0
$


# **Exercise 3**

**Exercise 3 part 1**

Let us assume that the amount of product(in kg) P and Q going to be produced in september 2021 are x and y respectively. So, the total cost associated with the production will be $145x + 280y$. For minimizing this cost the required linear program is as follows \\
$
minimize \ \ 145x+280y \\
sub \ to \ constraints \\
 x+y>= 110 \ (As\ combined\ production\ of\ P\ and\ Q\ should\ be\ at\ least\ 110) \\
 x>=84 \ (for\ satisfying\ major\ customer) \\
 y>=80 \ (for\ satisfying\ major\ customer) \\
 1.15x+0.92y<=430 \ (Time\ constraint) \\
 x,y>=0 \ (As\ amount\ of \ items\ produced\ cannot\ be\ negative)
$  

In [16]:
model3 = ConcreteModel()
model3.x = Var(domain=Reals)
model3.y = Var(domain=Reals)
model3.cost = Objective(expr = 145*model3.x + 280*model3.y, sense= minimize)

In [17]:
model3.constr1 = Constraint(expr = model3.x + model3.y >= 110)
model3.constr2 = Constraint(expr = model3.x >= 84)
model3.constr3 = Constraint(expr = model3.y >= 80)
model3.constr4 = Constraint(expr = 1.15*model3.x + 0.92*model3.y <= 430)
model3.constr5 = Constraint(expr = model3.x >= 0)
model3.constr6 = Constraint(expr = model3.y >= 0)

In [18]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 34580.0
  Upper bound: 34580.0
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 3
  Number of nonzeros: 9
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011408329010009766
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [19]:
print('\nCost = ', model3.cost())

print('\nDecision Variables')
print('x = ', model3.x())
print('y = ', model3.y())

print('\nConstraints')
print('Constraint 1  = ', model3.constr1())
print('Constraint 2  = ', model3.constr2())
print('Constraint 3  = ', model3.constr3())
print('Constraint 4  = ', model3.constr4())
print('Constraint 5  = ', model3.constr5())
print('Constraint 6  = ', model3.constr6())


Cost =  34580.0

Decision Variables
x =  84.0
y =  80.0

Constraints
Constraint 1  =  164.0
Constraint 2  =  84.0
Constraint 3  =  80.0
Constraint 4  =  170.2
Constraint 5  =  84.0
Constraint 6  =  80.0


**Exercise 3 part 3** \\
The optimal value is ** 34580**
and the value of decision variables are as follows \\
$x$ = 84, $y$ =80

Activities of all constraints respectively \\
Value of LHS followed by whether equality holds or not.

1.   164       , Equality does not holds
2.   84        , Equality holds
3.   80, Equality holds
4.   170.2    , Equality does not holds
5.   84     , Equality does not holds
6.   80      , Equality does not holds

**Exercise 3 part 4** \\
Due to a sudden demand of 102kg of product P and 117kg of poduct Q only constraint 2 and 3 will change. So the new problem become //
$
minimize \ \ 145x+280y \\
sub \ to \ constraints \\
 x+y>= 110 \ (As\ combined\ production\ of\ P\ and\ Q\ should\ be\ at\ least\ 110) \\
 x>=186 \ (for\ satisfying\ major\ customer) \\
 y>=197 \ (for\ satisfying\ major\ customer) \\
 1.15x+0.92y<=430 \ (Time\ constraint) \\
 x,y>=0 \ (As\ amount\ of \ items\ produced\ cannot\ be\ negative)
$  

In [20]:
model4 = ConcreteModel()
model4.x = Var(domain=Reals)
model4.y = Var(domain=Reals)
model4.cost = Objective(expr = 145*model4.x + 280*model4.y, sense= minimize)

In [21]:
model4.constr1 = Constraint(expr = model4.x + model4.y >= 110)
model4.constr2 = Constraint(expr = model4.x >= 186)
model4.constr3 = Constraint(expr = model4.y >= 197)
model4.constr4 = Constraint(expr = 1.15*model4.x + 0.92*model4.y <= 430)
model4.constr5 = Constraint(expr = model4.x >= 0)
model4.constr6 = Constraint(expr = model4.y >= 0)

In [22]:
!apt-get install -y -qq glpk-utils

In [23]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model4).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 82130.0
  Upper bound: 82130.0
  Number of objectives: 1
  Number of constraints: 7
  Number of variables: 3
  Number of nonzeros: 9
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.012555122375488281
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [24]:
print('\nCost = ', model4.cost())

print('\nDecision Variables')
print('x = ', model4.x())
print('y = ', model4.y())

print('\nConstraints')
print('Constraint 1  = ', model4.constr1())
print('Constraint 2  = ', model4.constr2())
print('Constraint 3  = ', model4.constr3())
print('Constraint 4  = ', model4.constr4())
print('Constraint 5  = ', model4.constr5())
print('Constraint 6  = ', model4.constr6())


Cost =  82130.0

Decision Variables
x =  186.0
y =  197.0

Constraints
Constraint 1  =  383.0
Constraint 2  =  186.0
Constraint 3  =  197.0
Constraint 4  =  395.14
Constraint 5  =  186.0
Constraint 6  =  197.0


**Exercise 3 part 6** \\
The optimal value is ** 82130**
and the value of decision variables are as follows \\
$x$ = 186
, $y$ =197

Activities of all constraints respectively \\
Value of LHS followed by whether equality holds or not.

1.   383       , Equality does not holds
2.   186        , Equality holds
3.   197, Equality holds
4.   395.14    , Equality does not holds
5.   186     , Equality does not holds
6.   197      , Equality does not holds

Here the increased demand of P and Q just got added to the optimal values of P and Q obtained in part 3 and hence corrospondingly increased the optimal value of cost. \\
I would tell Mapple industries that the new requirement has increased the cost linearly. And they should try to maximize the profit instead of minimizing the cost of production.